Implement function multi versioning in sysimg

yuyichao opened the pull request:
Current TODO:

* [x] Implement more flexible cloning check

* [x] Handle "PLT" cloning

* [x] Generate dispatch data

* [x] Update sysimg loading

* [x] Fix `code_native`

* [x] ? Figure out what's wrong with AVX on x86.

-------
Original post

This is a WIP PR implementing function multi versioning in sysimg. It's still WIP but I'd like to get some feedback.

Current limitation

* x86 only.

    AArch64 support is easy to add though harder to test.
    ARM with new enough glibc can also support this.

* AVX2 only.

    Because this is what I care about the most. The pass needs to be improved to handle multiple versions instead of only two.

* Improve cloning heuristics.

    This would be needed for KNL, which might need cloning the whole code section. It shouldn't be that bad. Current heuristic clones < 10% of the functions. The code section is still much smaller than the rodata section FWIW....

* User interface

    `-C` option should probably be extended to support this. We might want to parse it ourselves instead of letting LLVM do it since we'll need to emit feature detection code and also sort them.
    This should also replace/extend the current cpuid logic in `dump.c`

* Make JIT code use native features too.

What works.

* Sysimg generation finishes without crash

* With `JULIA_CPU_TARGET=x86-64` the sysimg correctly generates code for avx2 + fma and gets loaded correctly.

* `code_native` works automatically :tada:

* No dispatch overhead between cloned functions. (i.e. a direct call is generated from `test_clone_g` to `test_clone_f` below. These two functions will obviously be removed in the final version)

tkelman commented:
I agree with your choice of AVX2 giving the most benefit as a cutoff, if implementing 2 choices of instruction sets is much simpler to do than 3 or more. Haswell was introduced in 2013 (though are there any low-end Haswell or newer processors that have AVX2 disabled?). For Sandy Bridge (2011) and Ivy Bridge (2012) it might be useful to have AVX, but maybe not worth it for now. How does LLVM handle the operating system software needing to be new enough for AVX? Does it work correctly if the system the binaries get built on is very old?

yuyichao commented:
Implementing more than 2 is not hard. Just want to get the necessary LLVM pieces together first...

I don't think there's anything too much needed from LLVM, though we might want to disable avx/avx2 for old LLVM versions (if we don't drop them first) since we currently do that for some llvm versions https://github.com/JuliaLang/julia/blob/b1f668dfa97e44bbaec4ae1d5796a3f7f422a339/src/codegen.cpp#L7180

The multiversioning implemented here is purely an optimization and doesn't change external interface. (That'll be future work.) So there shouldn't be any requirement on system libs. LLVM seems to be pretty good at avoiding AVX - SSE transition panelty so calling SSE library from AVX code shouldn't cause performance issue, though it's another story on KNL, which is why KNL support will need to clone almost every functions.......

As for the default and binary release. I think the binary release can use sse2(default),avx,avx2+fma on x86_64. Not sure if we need to change default build. We can also include avx512 and knl if people are fine with the additional code size and if we actually have users.

tkelman commented:
There are no avx512 kernels in openblas yet (https://github.com/xianyi/OpenBLAS/issues/991) so I suspect most people on KNL are going to want source builds anyway. The effect on code size will be interesting to measure.

closes #14995 and could maybe replace some of #19486 ?

yuyichao commented:
> no avx512 kernels in openblas

K.... I guess when they have that it'll be a good time to revisite this. There'll probably also be Xeon's with AVX512 at that point and it'll probably make sense to enable avx512 support without KNL support (the two have very different cost model, avx512-skylake will likely not need cloning all functions).

> closes #14995 and could maybe replace some of #19486 ?

Yeah, being able to replace them is essentially the clean up work.

Keno commented:
> K.... I guess when they have that it'll be a good time to revisite this. There'll probably also be Xeon's with AVX512 at that point and it'll probably make sense to enable avx512 support without KNL support (the two have very different cost model, avx512-skylake will likely not need cloning all functions).

You can get one on Google Cloud right now.

yuyichao committed:
Remove code that are not needed/will not make sense with function multi-versioning and
improved cpu selection/detection.

* Remove cpuid specific binary

    To be replaced by function multi-versioning

* Remove old CPU target handling in sysimg

    The feature check is not complete since X86 features bits
    are also in many other CPUID results, including ones that
    we are actually interested in like AVX2.

* Add invalid CPU name and help handling in codegen initialization.

    Update `--cpu-target` test
    since target mismatch will not raise an error anymore.
    CPU names unknown to LLVM would.

yuyichao committed:
Move CPU feature detection functions to processor.cpp

Use C++ instead of C since there will be a lot of string handling and indirect interaction
with LLVM in this file. The meta-programing capability is also pretty useful.

tkelman commented on src/processor.cpp:
```diff
@@ -0,0 +1,164 @@
+#include "julia.h"
+#include "julia_internal.h"
+
+extern "C" {
```
am I missing something, or why have the file be `.cpp` instead of `.c` ?

yuyichao replied:
There'll be string manipulation functions here for parsing the command line args, parsing `/proc/cpuinfo`, generating LLVM feature string and I don't really want to do that in C (especially not for the output part). Basically I want to put the processor detection logic together instead of in 3+ files. This include the codegen initialization code at https://github.com/JuliaLang/julia/blob/3d5685a851af1068e9c8018d588f762a6188582d/src/codegen.cpp#L7089

tkelman replied:
ah okay. there won't be a C or C++ API somewhere in llvm or libuv for whatever you plan on using `/proc/cpuinfo` for?

yuyichao replied:
LLVM does have it but it's not exposing all the details (e.g. I find no way to access what features is enabled by a cpu, which may or may not be needed). More importantly we should be able to load the multi-versioning sysimg with LLVM disabled (or we need a `antiprocessor.c`............)

OpenBLAS also has it, though obviously not exposed in a useful way.

I'm not aware of such code in libuv.....

yuyichao committed:
Add X86 CPU name and feature detection

Also implement the API needed by other part of the runtime,
especially LLVM JIT, cloning pass and sysimg initialization.

Most of the detection (both name and features) code is copied from LLVM.
However, it's very hard for us to use LLVM because there are informations we need that
is not provided by LLVM (some of which is non-trivial to expose in a target independent way),
including,

1. Feature dependencies
2. Features related to a CPU name
3. Feature name validation (needed when we expose this to julia code)
4. Effect on ABI (vector register size)
5. Serialization format
6. Cloning heuristic

Additionally, the detection code itself is only a small part of the code for each arch
and we need to support multiple LLVM versions so copying the LLVM code shouldn't cause too
much problem by itself.

vtjnash commented on base/sysimg.jl:
```diff
@@ -400,6 +400,24 @@ end
+    return s
+end
+
+test_clone_g(Float64[], 1)
```
maybe hide these in `module Test`?

yuyichao replied:
This will be gone in the final version. It's just here for testing. It's pretty hard to actually come up with a useful test case that handles all machines the test can run on....

vtjnash replied:
very true. Another nice one to add support for with this infrastructure is `@llvm.fma.f64` (e.g. move the https://github.com/JuliaLang/julia/blob/a204bfdb4c2968ad5184c6e75451b0060851bf3f/base/floatfuncs.jl#L222 hack into the compiler)

vtjnash commented on src/llvm-mv.cpp:
```diff
@@ -0,0 +1,296 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+// Function multi-versioning
+#define DEBUG_TYPE "julia_mv"
```
let's actually call this file `multiversioning` or `ifunc`, mv is too cryptic

tkelman replied:
ifunc is pretty jargony, +1 for multiversioning

yuyichao replied:
I was expecting comments on this. Yeah, this will be renamed... `mv` was just simple to type....

vtjnash commented on src/processor.cpp:
```diff
@@ -0,0 +1,164 @@
+
+// Cache of information recovered from jl_cpuid.
+// In a multithreaded environment, there will be races on subnormal_flags,
+// but they are harmless idempotent races.  If we ever embrace C11, then
```
`c++11`, and don't we already require that?

yuyichao replied:
We do. And I've already replaced this with a global variable with dynamic initialization on my local version.

vtjnash commented on src/dump.c:
```diff
@@ -246,6 +247,28 @@ static void jl_load_sysimg_so(void)
+            if (nfunc && fptrs && fidxs) {
+                for (size_t i = 0; i < nfunc; i++) {
+                    size_t fi = fidxs[i];
+                    sysimg_fvars[fi] = fptrs[i];
```
I think I would rather leave this as a shared, read-only table and instead emit N tables (e.g. base table + (sparse?) overlay tables for each architecture). Otherwise we leak memory here, since we'll never want to reference it again after finishing loading. `dispatchf` seems still useful, to return the index of the appropriate overlay table to use (or, in the future, perhaps an ordered list of overlay tables).

yuyichao replied:
I haven't fully decide on this yet but in any case I'll probably make the big table read only.
I'll probably make a separate relocation table that's mutable and use that for dispatch instead and just add another loop in debuginfo.cpp. The memory usage can be further reduced using double indirection during the call but since how little the memory overhead actually is I prefer to keep it using only one indirection.

yuyichao committed:
Add AArch32 and AArch64 CPU name and feature detection code

Also implement internal runtime API.

The detection code avoid using `/proc/cpuinfo` whenever possible and should be much more
reliable than the one in LLVM. It also contains a much larger CPUID table to decode CPU names.

Compare to X86, the feature encoding/decoding is more complex due to the way LLVM takes
attributes. Certain information (arch version) also needs to be moved between name
and feature list.

tkelman commented on src/dump.c:
```diff
@@ -2590,13 +2585,16 @@ JL_DLLEXPORT void jl_preload_sysimg_so(const char *fname)
+    // TODO make sure the sysimg and the JIT agrees on the ABI.
+    // This shouldn't be a problem for any required C types on any platforms we support
+    // but could be a problem from optional types. In particular, we need to make sure
+    // the two agrees on the usuable register sizes so that functions that takes
```
usable

tkelman commented on test/cmdlineargs.jl:
```diff
@@ -61,15 +61,6 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes --startup-file=no`
     @test !success(`$exename --load`)
 
-    # --cpu-target
-    # NOTE: this test only holds true if image_file is a shared library.
```
why don't we want to run this test still?

vtjnash replied:
> I'm not aware of any method to ask LLVM if it is a supported target name at runtime

You mean `jl_TargetMachine->getMCSubtargetInfo()->isCPUStringValid()` 😜 

(added in https://llvm.org/svn/llvm-project/llvm/trunk@223147)

yuyichao replied:
Doesn't seem to be a way to do that without looking up the CPU first?

yuyichao commented on test/cmdlineargs.jl:
```diff
@@ -61,15 +61,6 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes --startup-file=no`
     @test !success(`$exename --load`)
 
-    # --cpu-target
-    # NOTE: this test only holds true if image_file is a shared library.
```
It won't make sense anymore. This was testing that a cpu target different from the one used to compile sysimg will throw an error but it will no longer be the case.

tkelman replied:
shouldn't an invalid setting for that option still error?

yuyichao replied:
No. It could be an CPU name that we don't recognize but llvm does so it shouldn't be blocked at this level and should be passed to LLVM. Unfortunately LLVM does not throw an error in that case and I'm not aware of any method to ask LLVM if it is a supported target name at runtime.

yuyichao commented:
Just a status update. This is still largely WIP but a few boring part are done.

* X86 arch and feature detection is complete. This can detect as many features as LLVM-svn (`sgx` is disabled for now since it's hard to do and AFAICT LLVM doesn't do it correctly either)
* ARM and AArch64 feature bits are done. CPU model detection have also gathered [enough data](https://github.com/yuyichao/explore/wiki/ARM-cores). Thanks to people who provide CPU info of their phones and especially @staticfloat who provides one (or two?) kyro CPU id that isn't found elsewhere.
* The `crc32c` change is a test for the new internal feature test API. It'll also be good to implement similar optimization for aarch64 especially since every aarch64 core I've seen so far supports crc32c.
* CI runs are to make sure that the new arch/cloning is working on x86. The only issue left there seems to be the segfault on linux 32bit. I am able to reproduce it locally on llvm 3.9.1 but not 4.0. Maybe I'll just disable avx on 32bit linux for llvm < 4.0. It's unlikely going to matter much anyway....

yuyichao commented:
So the 32bit linux segfault is really funny. LLVM 3.9 didn't realize that xmm8 is not available in 32bit mode so it wanted to emit

```asm
vcvtsi2sdl      %ecx, %xmm8, %xmm3
vmovsd  %xmm2, -8(%esi)
vmovsd  %xmm3, (%esi)
```

which it illegally encoded as

```asm
c5 bb 2a d9     vcvtsi2sdl      %ecx, %xmm8, %xmm3
c5 fb 11 56 f8  vmovsd  %xmm2, -8(%esi)
c5 fb 11 1e     vmovsd  %xmm3, (%esi)
```

However, the correct decoding of that is actually
```asm
c5 bb 2a d9 c5 fb       ldsl    -70919894(%ebx), %edi
11 56 f8        adcl    %edx, -8(%esi)
c5 fb 11 1e     vmovsd  %xmm3, (%esi)
```

Which modifies the data segment register and causes the next instruction (as well as almost any other memory ops) to segfault (since it has a, well, faulty segment....)

Seems to be fixed in 4.0 so disabling avx on LLVM < 4.0 seems to be the correct solution.....

yuyichao commented:
And for the reference the simplified code that I used to reproduce this is

```julia
A = reshape(1.0:4.0, 2, 2)
B = Array{Complex{Float64}}(2, 2)
B .= 1 + 2im

@noinline f() = error()
function _broadcast2!(B::AbstractArray, A, A2, iter1, iter2)
    for I2 in iter2
        for I1 in iter1
            s1, s2 = A.dims
            (((1 <= I1) & (I1 <= s1)) && ((1 <= I2) & (I2 <= s2))) || f()
            ele = A.parent[I1 + (I2 - 1) * 2]
            bele = ele + A2
            B[I1, I2] = bele
        end
    end
end
_broadcast2!(B, A, Complex(false, false), 1:2, 1:2)
```

It's not particularly reliable since it likely strongly depend on register pressure.

yuyichao commented:
The win32 failure looks like https://github.com/JuliaLang/julia/pull/22205 so I guess it needs a llvm binary update from @tkelman.

This still needs a few more days/weeks to get ready so no hurry.

tkelman commented:
when you're sure your llvm patches are done, split them into their own pr, makes it simpler

yuyichao commented:
I don't think there's any new LLVM patches in this PR. Is the patch in https://github.com/JuliaLang/julia/pull/22205 already built into the binary?

tkelman commented:
should be, since that was merged before https://github.com/JuliaLang/julia/pull/22356 was opened

yuyichao commented:
Interesting, so something about avx (or other features) is still not working on 32bit..... I'll probably worry about it later....

yuyichao committed:
Implement function multi versioning in sysimg

* Implementing function cloning pass.
* Hook up debug info lookup and sysimg loading to the processor initialization API.

yuyichao committed:
Implement fallback CPU detection and dispatch API

Fallback to LLVM for CPU name and feature detection.
Still allow cloning but is restricted to exact name matching dispatch.

yuyichao committed:
Use processor.cpp API in disasm.cpp

Remove ARM specific workaround and do not assume the CPU name contains all
information about the code.
This should now allow disassemble of code that's not required by the base ISA
or even not supported on the current CPU.
This is especially important on ARM, which does not disassemble many armv7 instructions previously.

yuyichao committed:
Use processor.cpp API to initialize LLVM JIT

This is currently ignoring the sysimg data.
Will be fixed when we have the actual cloning pass.

yuyichao commented:
All features are implemented. Still need some clean up and testing but otherwise ready for review again. Note that even though this is marked as 1.1, this is slightly breaking (the first commit removes a compile option, an exported C function for that feature, and allow `jl_option.cpu_target` to be `NULL`).

yuyichao commented:
32bit failure is avx register issue again.......

JeffBezanson commented:
Could you explain a bit more how dispatch works?

yuyichao commented:
> Could you explain a bit more how dispatch works?

That comment appearing on github frontpage out of context was a bit frightening... (Only the author and PR number was shown, not PR name.......)

I think most of what can be called dispatch are documented in code comments but which part of the dispatch are you talking about?

1. Deciding which target to pick.

    The logic is pretty simple and described here https://github.com/JuliaLang/julia/blob/97b4937a5c11c1243fb6b3233d9abec0afbe4b7c/src/processor.cpp#L20

2. How the functions in the sysimg call each other.

    Cloned functions always only have direct calls. Uncloned function calling possibly cloned functiion uses a GOT that's initialized when sysimg is loaded. The logic to decide this is mentioned in https://github.com/JuliaLang/julia/blob/97b4937a5c11c1243fb6b3233d9abec0afbe4b7c/src/llvm-multiversioning.cpp#L1051

3. How does the runtime load/initialize the sysimg.

    It do so by getting information about the cloning and relocations needed from a few exported variables, described in detail here https://github.com/JuliaLang/julia/blob/97b4937a5c11c1243fb6b3233d9abec0afbe4b7c/src/processor.h#L42

4. Use of function addresses in sysimg.

    This is mainly for the plt's, which is mainly simple function addresses but could have aggregates due to global merging on arm and aarch64 (I'd really like to disable it until this pass is run but can't findout a way...) This is breifly commented on here https://github.com/JuliaLang/julia/blob/97b4937a5c11c1243fb6b3233d9abec0afbe4b7c/src/llvm-multiversioning.cpp#L707

JeffBezanson commented:
Thanks.

How do you decide which targets to support in a system image? Is it just one target for the base ISA and another that includes whatever features the host supports?

yuyichao commented:
> How do you decide which targets to support in a system image?

This is specified by the user on the command line. For example, what I'm testing locally on x64 is, `generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)` (which is the one I would recommend for the generic binary, possibly with the addition of avx512 now that skylake-x is released). This string specifies three targets, separated by `;`,

1. `generic`

    This is the basic requirement for the ISA, (with the exception of equivalent to `pentium4`/`sse2` on x86_32)

2. `sandybridge,-xsaveopt,clone_all`

    This is the `sandybridge` target without the `xsaveopt` feature. This target will have all of the functions in the sysimg cloned. `-xsaveopt` makes this compatible with all AMD processors that support AVX ISA extension.

3. `haswell,-rdrnd,base(1)`.

    Similar to the one above, this is `haswell` minus `rdrnd` (`rdrnd` is unlikely to be useful for the JIT and it is disabled by RR so explicitly disabling it makes it usable under RR). This target is partially cloned and the functions not being cloned will be taken from the `base(1)`, which is the `sandybridge` target (0 base indexing here). This target is compatible to all AMD and Intel CPU's that support 3-operand FMA+AVX2 (haswell+, ryzen).

yuyichao commented:
In another word, the cloning needs to be enabled by default and therefore is not being tested on the CI yet. I've just finished all the testing locally (with many issues exposed mainly on ARM/AArch64....) so I'll try to enable cloning on the CI.

JeffBezanson commented:
Very nice! This is actually a really cool feature --- good candidate for a blog post at some point. It's like a general and reusable version of the architecture support in a library like openblas.

It's a bit unfortunate that so much code needs to be copied. Is there a path to getting some of it upstreamed eventually?

yuyichao commented:
> good candidate for a blog post at some point

Jameson had said the same. I don't think I'm very good at writing blog posts though.....
I should also give credit to GCC since this work was inspired by the `target_clone` attribute and [this bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80732) That makes me realize that the core part is pretty easy to implement.

> It's a bit unfortunate that so much code needs to be copied. Is there a path to getting some of it upstreamed eventually?

Actually not **that** much code duplication. Most of the code copied from LLVM are basically data. The most important information not exposed by LLVM is the complete supported/enabled feature set as well as a few properties for each attribute. ARM has a much more complete CPU id list that I gathered and a few more reliable CPU id and feature detection methods. Some of them can probably be upstreamed. The general issue with upstreaming these code though is that we actually need a pretty tight integration (we do many different kinds of query on these data) in order to drive the cloning pass so as long as the cloning pass is still our own, it's much easier to have full control on the raw data or we'll likely need quite a few platform dependent API's to get all the informations we need. We might even need to get other informations when we have a user visible API (that's the next step for this) about the targets and it'll be a pain if we have to wait for LLVM to add them. That said, these are just simple informations, e.g. how many bits the vector register has, does the feature contains fma, or what code can benefit from this feature (currently we only have math/simd), but they are detailed enough that I don't think it's future proof to let llvm expose enough raw data and there's enough of them that I'm not sure if they are interested in that many single user APIs, unless some other projects want to do the same of course. I also feel like they are not that interested in these API given the subtargets are not public API AFAICT.

The cloning pass itself is ~1000LOC and while it is generic enough to handle a lot of cases, it's still not good enough as a general purpose pass and relies strongly on our runtime for initialization so a major overhaul would be needed to upstream it.....

JeffBezanson commented:
The data tables would be the nicest thing to factor out actually, to some kind of external `libcpuinfo`. Of course there's no need to do this any time soon, but it would be nice to share maintenance of those details with other projects.

yuyichao committed:
Finish up function multiversioning support

* Enable test function multiversioning on the CI

  We can't do too much cloning on the CI before hitting the timeout or memory limit...
  Also avoid turning on cloning on circle CI since we seem to be very close to the memory limit.

* Add devdoc

vtjnash commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1463 @@
+    // sysimg means. Make it an error for now.
+    if (cmdline.size() > 1) {
+        jl_error("More than one command line CPU targets specified when"
+                 "not generating sysimg");
```
I think we call this "sysimage"

vtjnash replied:
Or, alternatively, "without a `--output-x` flag specified"

vtjnash commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1463 @@
+    else {
+        auto &targetarg = cmdline[0];
+        if (targetarg.en.flags & JL_TARGET_CLONE_ALL)
+            jl_error("`clone_all` feature specified when not generating sysimg.");
```
use straight quotes (“clone_all”)

vtjnash commented on src/processor_fallback.cpp:
```diff
@@ -0,0 +1,185 @@
+}
+
+static uint32_t sysimg_init_cb(const void *id)
+{
```
This function (and various others in this file), don't really seem to need to be duplicated and conditionally compiled.

yuyichao replied:
Most of these functions have more difference than the code they can share. Using a callback is certainly possible but that'll make the code much harder to read (so I only use it for a few functions that has a lot of shared logic like cmdline parsing and sysimg loading). A few internal API functions can be moved out but they are small enough that I think being able to find them in a file easily is more useful than reducing a few lines of code. Some of the API function implementations also need platform specific tweaking so either they can't be all moved out (making it hard to find code) or new conditional compilation need to be added in the function (which also decrease readability).

The feature comparison part in `sysimg_init_cb` is indeed one of the few good candidate to be separated out as a helper function though one of the most important condition `max_vector_size` needs to be a callback, making it quite a bit more ugly.... (And FWIW, it took a bit of an effort to twist the condition on different archs to all fit this code and I am still impressed it actually works out with the same code..............................)

vtjnash replied:
> Using a callback is certainly possible

This is C++, so you can also use interfaces or templates, if that helps.

yuyichao replied:
Yes, I know. But a large number of those makes the the code really hard to read......

vtjnash commented on src/processor.cpp:
```diff
@@ -0,0 +1,737 @@
+        int cnt = 0;
+        for (size_t i = 0; i < n; i++) {
+#ifdef __GNUC__
+            cnt += __builtin_popcount(eles[i]);
```
use the llvm names (e.g. `llvm::countPopulation(eles[i])` in this case), from `#include <llvm/Support/MathExtras.h>`

yuyichao commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1463 @@
+        return 0;
+    unsigned long val = 0;
+    unsigned long buff[2];
+    while (read(fd, buff, sizeof(buff)) == sizeof(buff)) {
```
This is not a real file so it should reading partial result is not a concern. The file has identical layout as the auxiliary value vector which is a `long` key, followed by a `long` value. The loop read each entry until finding the right key or the end.

yuyichao commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1463 @@
+
+static inline bool str_startswith(const std::string &str, const char *prefix)
+{
+    return strncmp(str.c_str(), prefix, strlen(prefix)) == 0;
```
I didn't want to link to LLVM but I guess it's fine to just link to libLLVMSupport?

yuyichao commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1463 @@
+    if (maxidx >= 0) {
+        list.erase(std::remove_if(list.begin(), list.end(), [&] (std::pair<uint32_t,CPUID> &ele) {
+                    int idx = find(ele.first);
+                    return idx != -1 && idx < maxidx;
```
Remove only the elements that are clearly the little.

vtjnash replied:
But why not just write `idx != maxidx`?

yuyichao replied:
Because I want to keep -1.

yuyichao commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1463 @@
+    }
+}
+
+const auto host_cpu = [] {
```
I guess just adding `static` should work.

yuyichao commented:
Most of the comments (not the lazy initialization of `host_cpu` yet) addressed and I also added devdoc.

It turns out that avx512 is unusable pre-llvm 5.0 (at least not without https://github.com/JuliaLang/julia/issues/22414) so it is forced off on LLVM 3.9 and 4.0 for now.

yuyichao commented:
... Try again disabling avx512 on old llvm versions....

`host_cpu` is also lazily initialized now with some other clean ups...

yuyichao commented:
Also at least macOS travis actually passed that point https://travis-ci.org/JuliaLang/julia/jobs/280037275#L4950 with ~53 seconds to spare... so things seem to be working ias expected...

vtjnash commented:
> it already takes ~5min (mostly LLVM time) to optimize the sysimg on master

You're welcome to revive https://github.com/JuliaLang/julia/pull/16059. Most of it was already merged into master, just not the separation between creating (`--output-ji`) and compiling `(--output-o`) the sysimg.

yuyichao commented:
All CI failures are https://github.com/JuliaLang/julia/issues/23796 (It appears on appveyor too since avx was previously disabled)

yuyichao commented:
So this actually pass all CI now =) so I don't think there's anything blocking anymore.

A few other notes,

1. This obviously has an effect on the compile time and memory consumption but only when cloning is enabled so normal users compiling for native (by default) shouldn't notice a difference.

2. With this patch, generic binary will automatically pick up native features at runtime since I don't think the sysimg is currently using any features that breaks the runtime ABI with any x86 extensions (i.e. we don't have functions explicitly taking vector as argument or returning it as results). This can certainly expose more LLVM bugs (as shown when testing cloning on the CI) but I don't think that's a blocking issue for this PR. Also, all the ones I've found are worked around.

3. I'm not sure what binary we'll provide (i.e. what do we care about/what hardware people actually use) after this but here's my recommendation:

    * On X86_64

        Without LLVM 5.0 AVX 512 is unusable ( :cry: ) so I think `generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)` should be good enough. `generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1);skylake-avx512,-rdrnd,base(1)` can be a good choice when we upgrade to LLVM 5.0+.

    * On X86_32

        Without LLVM 5.0 AVX is unusable ( :cry: ) so I `generic;sandybridge,-xsaveopt,clone_all` should be good enough. Depending on whether we care about avx2/fma/avx512 on 32bit x86, we can use either one of the x64 target list on x86_32 after switching to LLVM 5.0+.

    * On AArch32 (armv7-a)

        I'm not sure about the user distribution but the useful features are likely vfp4 (for fma) and neon (for simd). So maybe `armv7-a;armv7-a,neon;armv7-a;armv7-a,neon,vfp4` would be good.

    * On AArch64

        The base architecture currently have almost all the important/interesting features since it's still very young so there isn't a lot of need for cloning. Just `generic` is probably good enough. (falkor or thunderx2 optimization would be interesting when the hardware are more widely available by the end of this year/early next year)

yuyichao commented:
Will merge before/during this weekend if there's no objections. I'll probably open an issue on the buildbot repo about changing the generic binary target since I don't think we'll change the default one.

vtjnash commented:
Should we trigger nanosoldier `runbenchmarks(ALL, vs=":master")`?

fredrikekre commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21849#issuecomment-332868597) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/53d4d6a9759c90138d36d5310b1f71dcfb0586d5/fb6921c_vs_d1b6f78/report.md). cc @ararslan

yuyichao commented:
The benchmark probably had some merge issue. I don't think those differences are real.....

vtjnash commented:
> Will merge before/during this weekend if there's no objections

Not really an objection, but I would like to try to at least read through all of the code changes. At my usual rate, +5000 lines will take awhile, so I've been trying to take an incremental approach. I'll try to prioritize getting through this soon though.

yuyichao commented:
Updated the cpu name check to throw an error if the JIT target is not known to LLVM. Also makes the behavior of `-C help` less annoy.

Now an invalid cpu name will always throw an error as long as LLVM is enabled (so the test is enabled unconditionally, given that we have many other tests that requires LLVM). `help` can be provided either as a feature or as a cpu name in which case the cpu name will be treated as `native` (the default) and the feature will be ignored while a single LLVM target help message will be printed. I still can't find a way to iterate through LLVM features, or we should be able to print a more julia-specific help message instead.....

yuyichao commented:
> +5000 lines will take awhile

Actually the only file that had major changes in the past month is `llvm-multiversioning.cpp`. (~1000LOC). The CPU feature detection/dispatch code has been the same since https://github.com/JuliaLang/julia/pull/21849#issuecomment-305950641 .

JeffBezanson commented:
OK to take some time with this --- it's not super urgent and it is a large amount of code.

yuyichao commented:
Any more review for this?

> it's not super urgent and it is a large amount of code.

True, though at the same time, almost none of the code will affect normal user (almost all of the code are for sysimg compilation) and the only way to test these code would be to build julia with different options on different machines, which would be much easier with this merged.

vtjnash commented on src/llvm-multiversioning.cpp:
```diff
@@ -0,0 +1,1067 @@
+                                                insert_before);
+        }
+        else {
+            jl_safe_printf("Unknown const use.");
```
Also could be a `GlobalIndirectSymbol`. Is the assert actually necessary?

yuyichao replied:
The assert is necessary since all of the uses needs to be fixed individually. (AFAICT LLVM doesn't have anything that does this recursive replace...).

This should handle everything that we emit and I don't think LLVM introduces any IFuncs or Aliases for functions automatically.

vtjnash replied:
We emit Aliases for functions, Why do we need to fix all uses?

yuyichao replied:
> We emit Aliases for functions

For `jl_extern_c` I guess? That's not fixable then.....

> Why do we need to fix all uses?

It's very bad to let the wrong version running.

yuyichao replied:
(`GlobalAlias` isn't the not-fixable part. `ExternalLinkage` is.)

vtjnash replied:
> It's very bad to let the wrong version running.
  
worse than calling `abort` (instead of, say, a warning)?

yuyichao replied:
Well, it may crash or give wrong result and even if that doesn't happen, I do think a compile time error is better than a mysterious runtime performance issue.

vtjnash commented on src/disasm.cpp:
```diff
@@ -670,7 +664,7 @@ static void jl_dump_asm_internal(
     // Set up Subtarget and Disassembler
     std::unique_ptr<MCSubtargetInfo>
-        STI(TheTarget->createMCSubtargetInfo(TripleName, MCPU, Features.getString()));
+        STI(TheTarget->createMCSubtargetInfo(TripleName, cpu, features));
```
I rather think we should just be using `"", ""` here (like the defaults for llvm-objdump).

Although if we have `object != NULL`, we could be very fancy, and use
`TheTriple = object->makeTriple(); features = object->getFeatures();`

yuyichao replied:
No that's what's causing https://bugs.llvm.org/show_bug.cgi?id=30525

yuyichao replied:
And I also don't want to rely on the feature set of the object since I think it should be allowed to embed instructions in the code that is not generally available as long as it's protected by feature checks.

vtjnash commented on src/processor.cpp:
```diff
@@ -0,0 +1,812 @@
+    const char *start = ext_features.c_str();
+    for (const char *p = start; *p; p++) {
+        if (*p == ',' || *p == '\0') {
+            features.emplace_back(start, p - start);
```
`*p` can't == `'\0'`

yuyichao replied:
Yes, this is a bug.

yuyichao replied:
Fixed.

yuyichao commented on src/processor_arm.cpp:
```diff
@@ -0,0 +1,1408 @@
+    auto dir = opendir("/sys/devices/system/cpu");
+    if (!dir)
+        return;
+    while (auto entry = readdir(dir)) {
```
Deprecated.

vtjnash commented:
Alright, one review comment open (in processor.cpp) and I think this is ready to merge.

yuyichao commented:
I've also got some local updates that tweaks the x86 handling a little. Will push later.

Handling of directly exported function is still missing, which should just mean that use of `jl_extern_c` can't have multiversioning. A fix with slight semantic change should be possible (essentially emit a trampoline in this pass, not in codegen directly since this pass might want to clone that function too....) That should be fairly independent of other changes can be fixed later.

vtjnash commented:
This broke the linux32 buildog tests: https://buildog.julialang.org/#/builders/41/builds/239
